### ISSUE TITLE
Discovery of heterogenous nodes / supporting multiple discovery methods

### ISSUE THREAD

**Issue by arjan** [mrluc/peerage#18]
I have an Elixir cluster spanning multiple k8s deployments; one deployment starts the Elixir VM with `-name a@${MY_POD_IP}`, the other as `-name b@${MY_POD_IP}`

The DNS discovery mechanism currently assumes all nodes have an identical `app_name` (the part before the @), so these deployments cannot find each other even though they should.

I see two possible solutions:

1) making the `app_name` config into a list so that it can try to connect to all `a@` and `b@` nodes on each discovered IP address, but this will of course result in half of the `Node.connect()` calls failing:

```
    config :peerage, via: Peerage.Via.Dns,
      dns_name: "myapp", # or k8s FQDN: "myapp.ns.svc.clust.local"
      app_name: ["a", "b"]
```

2) Allowing the `via` config to be a list of discovery methods instead of just one:
```
config :peerage, via: [
  {Peerage.Via.Dns, 
   dns_name: "domain-a.ns.svc.clust.local", 
   app_name: "a"},
  {Peerage.Via.Dns, 
   dns_name: "domain-b.ns.svc.clust.local", 
   app_name: "b"}
  ]
```

Personally I think #2 is the best. What do you think? I'll create a PR, I need this for a project.

**Comment by arjan** [comment]
Method nr 3: Come to think of it, given the statefulness of the Udp method, and the unstatefulness of the other methods, it might make most sense just to have the `Dns` provider be multi-app aware:

```
config :peerage, via: Peerage.Via.Dns, dns: [
  [dns_name: "domain-a.ns.svc.clust.local", app_name: "a"],
  [dns_name: "domain-b.ns.svc.clust.local", app_name: "b"]]

```

**Comment by mrluc** [comment]
Hi @arjan — that’s the scenario I have as well; usually with at least one app deployed as a headless service, and others that talk to the outside world via ingress or usual lb. If you just configure both a and b to use the ‘a’ app for discovery, and ‘a’ is a headless service, due to how erlang distribution works, they should all know about each other.

There is still value to be discussed in providing multiple names, not least because using one name for discovery is a single point of failure.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by mrluc** [comment]
(The docs for dns provider might need to be updated to not imply that the name be the app’s own name, as it shouldn’t need to be - rather, it’s the name you want to use for discovery).

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by arjan** [comment]
Well, it's the *node*'s name, right? The part without the FQDN? (`--name`)
The DNS provider generates a list of nodes based this app_name + '@' + IP address; but I have 2 distinct app names (a@ and b@) in my Elixir cluster; they basically distinguish the "role" they have. Or am I missing something in how Erlang distribution works…

**Comment by mrluc** [comment]
@arjan maybe - if there are 10 ‘a’ nodes, and 10 ‘b’ nodes, and the A service is headless so that lookup returns multiple ips instead of using a load balancer ... then yes, just pointing all A _and_ all B at the ips from A will result in them all knowing about each other.

The missing ingredient is, yes, erlang distribution - once you connect to a node it will also tell you about other nodes it knows about, and assuming they’re reachable you’ll get a ‘strongly connected’ cluster, ie, every node connected to every other.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by arjan** [comment]
Yes you're right it could work like that. But still the DNS provider generates wrong node names for all the 'b' nodes, causing those connects to fail; in the end it probably does not matter that much, just does not feel right.

**Comment by king-xt** [comment]
@arjan - that sounds like something we'd like to prevent; can you expand a bit on what sort of failures you're seeing when you mention:

> still the DNS provider generates wrong node names for all the 'b' nodes, causing those connects to fail

Ie, what node names are being generated? Is the k8s service pointed at for discovery 'headless'?

I don't think that the symptom you describe should ever *need* to happen; it could be possible depending on how you're using k8s, though, as k8s has a lot of surface area! One way it could be possible that I can think of offhand is if (for whatever reason) some 'A' and 'B' k8s Deployments were both backing a single k8s Service, but there are probably other ways as well.

But assuming a setup like the following: given apps A and B, 10 A and 10 B nodes are deployed behind A-service and B-service, with A-service being 'headless' (`clusterIP: None`), and both deployed apps having config like:

```elixir
config :peerage, via: Peerage.Via.Dns,
      dns_name: "a-service", # or k8s FQDN: "a-service.ns.svc.clust.local"
      app_name: "a"
```

Then the DNS provider would only return new node names based on the IPs returned from looking up the 'a-service' name, of the `a@X.X.X.X` variety, and peerage wouldn't be trying to connect to any 'b' s at all in that scenario! (Rather, normal Erlang distribution will take care of strongly connecting the cluster). So I don't believe any of those connections should fail (except, of course, for the self-connects of the nodes backing the 'a-service', which is an issue in latest OTP but fixed in #17).

**Comment by mrluc** [comment]
@arjan - that sounds like something we'd like to prevent; can you expand a bit on what sort of failures you're seeing when you mention:

> still the DNS provider generates wrong node names for all the 'b' nodes, causing those connects to fail

Ie, what node names are being generated? Is the k8s service pointed at for discovery 'headless'?

I don't think that the symptom you describe should ever *need* to happen; it could be possible depending on how you're using k8s, though, as k8s has a lot of surface area! One way it could be possible that I can think of offhand is if (for whatever reason) some 'A' and 'B' k8s Deployments were both backing a single k8s Service, but there are probably other ways as well.

But assuming a setup like the following: given apps A and B, 10 A and 10 B nodes are deployed behind A-service and B-service, with A-service being 'headless' (`clusterIP: None`), and both deployed apps having config like:

```elixir
config :peerage, via: Peerage.Via.Dns,
      dns_name: "a-service", # or k8s FQDN: "a-service.ns.svc.clust.local"
      app_name: "a"
```

Then the DNS provider would only return new node names based on the IPs returned from looking up the 'a-service' name, of the `a@X.X.X.X` variety, and peerage wouldn't be trying to connect to any 'b' s at all in that scenario! (Rather, normal Erlang distribution will take care of strongly connecting the cluster). So I don't believe any of those connections should fail (except, of course, for the self-connects of the nodes backing the 'a-service', which is an issue in latest OTP but fixed in #17).

**Comment by arjan** [comment]
I was using a headless service, with both deployments of a and b exposed in that single headless service. which would yield a list of all IP addresses of all the nodes, instead of just a. 

Using just a in the headless service indeed is enough, given that the Erlang distribution takes care of the fully-connectedness of the cluster for us.

So I guess this issue can be closed for now :-) Thanks a lot for the clarification!
